### ISSUE TITLE
Need shortening of unsettled promise chains

### ISSUE THREAD

**Issue by erights** [Agoric/Agoric#4018]
Currently, unsettled promise chains that go through intermediate vats do not get shortened. This creates several problems, at least:
   * intervat and interchain promise cycles never get collapsed into a local cycle, and thereby never become detectable. Thus, messages sent into such loops keep cycling forever, interleaved with productive events. Thus, this does not prevent actual progress, but becomes a form of resource exhaustion attack slowing down other progress.
   * If an intermediate vat gets killed, the promise chain that should have been shortened instead breaks. Had the chain been shorted (and once we do implement shortening) the promise chain would survive the death of the intermediate vat, which will be an observable semantic difference.
   * We continue to pay the cost of the intermediate hops. When distributed across multiple chains this can get very expensive.

To implement unsettled promise chain shortening, we need to implement something like E's `__whenMoreResolved` protocol or what Midori did (@dtribble please fill in). To do so, we need to detect when a promise is forwarded to a handled promise. When a handled promise is forwarded to a handled promise, our shim can detect that, and could provide the necessary hooks to implement its part of a shortening protocol. However, without platform support, the shim *cannot* detected when a regular promise is forwarded to a handled promise. This happens implicitly surprisingly often, for example
```js
const foo = async () => handledPromise
const hp = foo();
```
Here, the `foo` function looks like it is returning a handled promise. The `foo` function does not even have an `await` in it, so it looks like it returns the `handledPromise` immediately. However, the `hp` variable is bound to a regular promise that has been implicitly forwarded to the handled promise. A platform implementation of the eventual-send proposal would treat this correctly. The shim cannot. Thus, if a forwarding chain involves such a regular promise under the shim, this has two consequences, which kinda compensate for each other:
   * If an intervat unsettled promise chain involves `fp`, under any plausible enhancement of the shim, it is still impossible to shorten this part of the chain.
   * Any message eventually sent into this chain stalls at `fp`, breaking pipelining and changing message order (problems) but also preventing the message from infinitely cycling around the loop (a surprising silver lining).

Thanks to Informal Systems for focusing our attention on this. Attn @jessysaurusrex @warner @michaelfig @dtribble

**Comment by erights** [comment]
Because of the silver lining above, even under the shim, it should be possible for us to implement enough shortening that we can prevent infinitely cycling messages. The remaining cycles we cannot detect will also stall messages rather than forwarding them.

**Comment by erights** [comment]
See
https://github.com/Agoric/agoric-sdk/issues/3923
https://github.com/Agoric/agoric-sdk/issues/3702 attn @danwt
https://github.com/Agoric/agoric-sdk/issues/1636
https://github.com/Agoric/agoric-sdk/issues/1354

@michaelfig Do we have a bug somewhere explaining why it is impossible for the shim to implement the full proposal? I recall that we did, including example code. But I can't find it.

**Comment by michaelfig** [comment]
> @michaelfig Do we have a bug somewhere explaining why it is impossible for the shim to implement the full proposal? I recall that we did, including example code. But I can't find it.

https://github.com/tc39/proposal-eventual-send#platform-support

**Comment by erights** [comment]
Hi @mhofman @michaelfig @warner @dckc   just added y'all. Something to take into account for pseudo-promise design? For ocapn? (attn @cwebber )

**Comment by mhofman** [comment]
It is very much my plan to support shortening in the pseudo-promise design, all the way back into the kernel through vats. That will also make it possible to detect promise cycles.

Of course shortening over ocapn is another story. As far as I understand it can just be expressed as a promise being resolved with another promise?

**Comment by erights** [comment]
There are some tricky issues with preserving message order of messages already in the air while shortening. E got these right. I think Midori got them more right (cc @dtribble ). When we shorten in the protocol, we need to be careful about this.

**Comment by mhofman** [comment]
We decided to only guarantee message order of messages sent to the same target by a given source. The question at hand is then whether a message sent to a promise resolving to an object and a message sent directly to that object have any ordering guarantees when sent by the same vat. I am somewhat of the opinion they do not. By extension I do not believe that messages sent by a vat to different promises resolving to the same object should have any implicit ordering guarantees.

**Comment by erights** [comment]
This is probably not the simplest scenario for illustrating this case, but it is the one in my head that's clearest to me as I write this.

Alice in vatA on chainA:
```js
const { promise: p1, resolve: r1 } = makePromiseKit();
E(bob).foo(p1);
```
Bob, in vatB on chainB, on receiving the `foo(p)` message, periodically sends a `bar(count)` message to `p` with an increasing count.
```js
const bob = Far('Bob', {
  foo: async p => {
    let count = 0;
    const notifierP = E(timerService).makeNotifier(delay, interval);
    const notifier = makeNotifier(notifierP);
    for await (const _ of notifier) {
      E(p).bar(count);
      count += 1;
    }
  },
});
```
Alice, sometime later, asks Carol in vatC on chainC for a reference to Charles, also in vatC
```js
const p2 = E(carol).charles();
r1(p2);
```
Let's assume that the `charles()` message to Carol arrives much later.
Carol and Charles, in vatC
```js
const charles = Far('Charles', {
  bar: count => console.log(count);
});
const carol = Far('Carol', {
  charles: () => charles,
});
```
Assume no other relevant activity anywhere in the system. Assume no partitions, crashes, or upgrades, or any other such trauma. In that case, I would expect Charles to print counts on vatC's console in strictly increasing order starting at `0` and not missing any natural numbers. This is motivated only by extending point-to-point FIFO intuitions to messages sent into unresolved promises. It does not rely on the intuitions from stronger orders like E-Order or Midori's (@dtribble 'a) Flow order. After all, Bob the object is completely unaware that his `p` promise has been rerouted from pointing towards vatA to pointing towards vatC.

Before `r1(p2)`, the `bar(count)` messages are being pipelined in order from vatB to vatA. Let's say that `bar(3)` is already in the air before Alice does `r1(p2)`. After Alice does `r1(p2)` but before Carol receives `charles()`, by unresolved promise chain shortening, Bob's remote promise `p` gets rerouted from pointing towards `r1` in vatA to pointing towards the decider of `p2` in vatC. (We can think of "the decider of `p2`" as an implicit `r2` carried in the outgoing `charles()` message.)

After Bob's `p` is rerouted, but while `bar(3)` is still in the air going from vatB to vatA, Bob sends `bar(4)` to his `p`. Now, `bar(4)` is sent from vatB directly to vatC while `bar(3)` is still in the air from vatB to vatA.

Both `bar(3)` and `bar(4)` will eventually be delivered to Charles. However, without extra mechanism (such as implemented in E's captp and in Midori), they may arrive at Charles out or order.

**Comment by erights** [comment]
> We decided to only guarantee message order of messages sent to the same target by a given source.

From Bob's perspective, all the messages were sent to the same `p` target by the same `Bob` source.

> The question at hand is then whether a message sent to a promise resolving to an object and a message sent directly to that object have any ordering guarantees when sent by the same vat. I am somewhat of the opinion they do not. By extension I do not believe that messages sent by a vat to different promises resolving to the same object should have any implicit ordering guarantees.

From Bob's perspective, all `bar(count)` messages were sent to the same `p` promise.

**Comment by mhofman** [comment]
Yes, I agree in this case the sender has been sending all messages to the same target from its point of view, and any pipelining and subsequent resolution would need to guarantee that expected order.

I'd imagine this problem is not unique to promise shortening as even messages sent to p1 and queued in vatA before resolution/forwarding would need to be delivered before any messages sent to p1 after it.

I am not sure how to implement this ordering guarantee in the protocol, especially in the case of mutually suspicious chains. I guess that's where pipelining gets complicated.

**Comment by mhofman** [comment]
I think it would be ok for vatB to cooperate with vatA on this ordering guarantee since vatB has been trusting vatA as the decider, but I'm not sure what that could look like. I dont think vatA and vatB can/should rely on vatC to respect any ordering A and B might have agreed upon.
